### PR TITLE
Hive 23100

### PR DESCRIPTION
**Pull request by jcamachor**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveCalciteUtil.java [hunk 1]
```diff
@@ -1062,6 +1060,25 @@ public RexNode apply(RelDataTypeField input) {
...
   }
 
+  public static boolean isConstant(RexNode expr) {
+    if (expr instanceof RexCall) {
+      RexCall call = (RexCall) expr;
+      if (call.getOperator() == SqlStdOperatorTable.ROW ||
+          call.getOperator() == SqlStdOperatorTable.ARRAY_VALUE_CONSTRUCTOR ||
+          call.getOperator() == SqlStdOperatorTable.MAP_VALUE_CONSTRUCTOR) {
+        // We check all operands
+        for (RexNode node : ((RexCall) expr).getOperands()) {
```
Comment (scarlin-cloudera) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveCalciteUtil.java:1070]: call.getOperands()

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveCalciteUtil.java [hunk 2]
```diff
@@ -1062,6 +1060,25 @@ public RexNode apply(RelDataTypeField input) {
     return HiveProject.create(input, copyInputRefs, null);
   }
 
+  public static boolean isConstant(RexNode expr) {
+    if (expr instanceof RexCall) {
+      RexCall call = (RexCall) expr;
+      if (call.getOperator() == SqlStdOperatorTable.ROW ||
+          call.getOperator() == SqlStdOperatorTable.ARRAY_VALUE_CONSTRUCTOR ||
+          call.getOperator() == SqlStdOperatorTable.MAP_VALUE_CONSTRUCTOR) {
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveCalciteUtil.java:1068]: I think support to recognize `SqlMultisetValueConstructor` would need to be added in other places too in any case... We can leave it more explicit for the time being and generalize later on if needed. Fwiw this call is also very specific to Hive and how it handles the initialization of UDFs with parameters that are constant; otherwise, we would not need it.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/ExprNodeConverter.java [hunk 3]
```diff
@@ -227,160 +228,165 @@ public ExprNodeDesc visitCall(RexCall call) {
...
-      case INTERVAL_HOUR:
-      case INTERVAL_HOUR_MINUTE:
-      case INTERVAL_HOUR_SECOND:
-      case INTERVAL_MINUTE:
-      case INTERVAL_MINUTE_SECOND:
-      case INTERVAL_SECOND:
-        return new ExprNodeConstantDesc(TypeInfoFactory.intervalDayTimeTypeInfo, null);
-      default:
-        return new ExprNodeConstantDesc(TypeInfoFactory.voidTypeInfo, null);
+        case BOOLEAN:
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/ExprNodeConverter.java:239]: Good catch! It was IntelliJ auto-formatting.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/stats/HiveRelMdSize.java [hunk 4]
```diff
@@ -183,4 +185,50 @@ public static Double averageTypeSize(RelDataType type) {
     }
   }
 
+  public static double typeSize(RelDataType type, Comparable value) {
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/stats/HiveRelMdSize.java:188]: The method was taken from Calcite's {{RelMdSize}} as is. Interestingly, I realized that although constructor for super class is protected, I could reuse it since HiveRelMdSize is extending the class. As a bonus, I could remove `averageTypeSize` too that was there for legacy reasons and use the one in `RelMdSize`.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/stats/HiveRelMdSize.java [hunk 5]
```diff
@@ -183,4 +185,50 @@ public static Double averageTypeSize(RelDataType type) {
...
+      case INTERVAL_SECOND:
+        return 8d;
+      case BINARY:
+      case VARBINARY:
+        return ((ByteString) value).length();
+      case CHAR:
+      case VARCHAR:
+        return ((NlsString) value).getValue().length() * BYTES_PER_CHARACTER;
+      default:
+        return 32;
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/stats/HiveRelMdSize.java:230]: Marking as solved since the method is removed.

**Comment by jcamachor** [comment]
Thanks for the feedback @scarlin-cloudera . I have pushed a new commit that addresses your comments.

Please @vineetgarg02 , @kgyrtkirk , take a look whenever you have the chance.

File: ql/src/test/results/clientpositive/llap/vector_between_in.q.out [hunk 6]
```diff
@@ -269,16 +269,16 @@ STAGE PLANS:
             Map Operator Tree:
                 TableScan
                   alias: decimal_date_test
-                  filterExpr: (cdecimal1) IN (2365.8945945946, 881.0135135135, -3367.6517567568) (type: boolean)
+                  filterExpr: ((cdecimal1) IN (2365.8945945946, -3367.6517567568) or (cdecimal1) IN (881.0135135135)) (type: boolean)
```
Comment (vineetgarg02) [inline ql/src/test/results/clientpositive/llap/vector_between_in.q.out:272]: This looks like regression where expressions are not further simplified like before. May be create a follow up jira?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/type/RexNodeExprFactory.java [hunk 7]
```diff
@@ -0,0 +1,1021 @@
...
+      return rexBuilder.makeLiteral(RexLiteral.value(literal), t, false);
+    }
+    throw new RuntimeException("Unsupported expression type: " + expr.getClass().getCanonicalName());
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected boolean canConvertCASEIntoCOALESCEFuncCallExpr(GenericUDF genericUDF, List<RexNode> inputs) {
```
Comment (kgyrtkirk) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/type/RexNodeExprFactory.java:835]: I don't really like this method...Ibelieve this functionality should live somewhere else instead (HivePointLookupOptimizer) ?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java [hunk 8]
```diff
@@ -223,7 +224,13 @@ public Double visitCall(RexCall call) {
           return null;
         }
         cost += operandCost;
-        Double size = HiveRelMdSize.averageTypeSize(operand.getType());
```
Comment (kgyrtkirk) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java:226]: I believe this is the change which have caused lot of operand orders to change
because this seems to be unrelated to HIVE-23100 to me - could it have it's own ticket?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/jdbc/JDBCExpandExpressionsRule.java [hunk 9]
```diff
@@ -196,24 +197,11 @@ private RexNode transformIntoOrAndClause(RexBuilder rexBuilder, RexCall expressi
...
-              valueExpressions.getOperands().get(j)));
-        }
-        if (conjuncts.size() > 1) {
-          disjuncts.add(rexBuilder.makeCall(
-              SqlStdOperatorTable.AND,
-              conjuncts));
-        } else {
-          disjuncts.add(conjuncts.get(0));
-        }
+      final List<RexNode> disjuncts = HiveCalciteUtil.transformIntoOrAndClause(
```
Comment (kgyrtkirk) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/jdbc/JDBCExpandExpressionsRule.java:200]: I believe this call should get the "expression" as argument not it's operands - that way the HiveCalciteUtil.tIOA method would have better contract - it seems like this callsite removes cases when there is a struct on the lefthandside - I believe that decision should be made by the utility method
...and then the OR will be created inside that method as well...it will be more straight forward

File: ql/src/java/org/apache/hadoop/hive/ql/parse/type/HiveFunctionHelper.java [hunk 10]
```diff
@@ -0,0 +1,549 @@
...
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public RexNode getExpression(String functionText, FunctionInfo fi,
+      List<RexNode> inputs, RelDataType returnType)
+      throws SemanticException {
+    // See if this is an explicit cast.
+    RexNode expr = RexNodeConverter.handleExplicitCast(
```
Comment (kgyrtkirk) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/type/HiveFunctionHelper.java:236]: I think it would be probably cleaner to create a RexNodeConverter; and pass a rexBuilder to it
after that :  remove the static from those methods - and the rexBuilder passed to every invocation of it...

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/RexNodeConverter.java [hunk 11]
```diff
@@ -936,14 +678,10 @@ protected RexNode convert(ExprNodeConstantDesc literal) throws CalciteSemanticEx
...
       final TimestampString tsLocalTZString;
-      if (value == null) {
-        tsLocalTZString = null;
-      } else {
-        Instant i = ((TimestampTZ)value).getZonedDateTime().toInstant();
-        tsLocalTZString = TimestampString
-            .fromMillisSinceEpoch(i.toEpochMilli())
-            .withNanos(i.getNano());
-      }
+      Instant i = ((TimestampTZ)value).getZonedDateTime().toInstant();
```
Comment (kgyrtkirk) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/RexNodeConverter.java:681]: I'm not sure; but the null check seemed to be needed here.
or...there can't be a "null" which is typed as TIMESTAMPLOCALTZ?

File: ql/src/test/results/clientpositive/llap/vector_between_in.q.out (hunk 6 above)
Comment (jcamachor) [reply ql/src/test/results/clientpositive/llap/vector_between_in.q.out:272]: Yes, you are right. This is a missed merging opportunity due to different type inferred for values (although they are all decimal, note that different precision,scale inferred for them). I have created https://issues.apache.org/jira/browse/HIVE-23222 to tackle this.

File: ql/src/test/org/apache/hadoop/hive/ql/plan/mapping/TestCounterMapping.java [hunk 12]
```diff
@@ -183,11 +183,11 @@ public void testBreakupAnd2() throws ParseException, CommandProcessorException {
     List<FilterOperator> fos = pm.getAll(FilterOperator.class);
     OpTreeSignature filterSig = pm.lookup(OpTreeSignature.class, fos.get(0));
     Object pred = filterSig.getSig().getSigMap().get("getPredicateString");
-    assertEquals("(u = 2) (type: boolean)", pred);
-
+    // TODO: Missed opportunity due to cast on constant preventing simplification.
+    // Fixed when we upgrade to Calcite 1.22.0
+    assertEquals("((u = 2) and (u) IN (1, 2, 3)) (type: boolean)", pred);
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/test/org/apache/hadoop/hive/ql/plan/mapping/TestCounterMapping.java:188]: Yes, when I saw this change the first time, I thought I would have to extend Calcite logic to simplify this. Then I realized it was simplifying it in 1.22. The culprit seems to be https://issues.apache.org/jira/browse/CALCITE-3525.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java (hunk 8 above)
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java:226]: Yeah... The reason why I have it in this patch is that a lot of q files were changing too without it. It seems the main reason is the subtle changes in the handling of char, varchar, and string which lead to different ordering.
I could definitely create a new issue for this, but given that it will not reduce the number of q file changes and that the method `typeValueSize` is part of Calcite, could we include it here?

File: ql/src/test/results/clientpositive/vector_case_when_2.q.out [hunk 13]
```diff
@@ -998,13 +998,13 @@ STAGE PLANS:
                 native: true
                 vectorizationSchemaColumns: [0:q548284:int, 1:ROW__ID:struct<writeid:bigint,bucketid:int,rowid:bigint>]
             Select Operator
-              expressions: q548284 (type: int), CASE WHEN ((q548284 = 4)) THEN (0.8) WHEN ((q548284 = 5)) THEN (1) ELSE (8) END (type: decimal(2,1))
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/test/results/clientpositive/vector_case_when_2.q.out:1001]: Good catch. It seems this is not needed indeed so we are missing an optimization opportunity here. I have created https://issues.apache.org/jira/browse/HIVE-23223.

File: ql/src/test/results/clientpositive/structin.q.out [hunk 14]
```diff
@@ -96,10 +96,10 @@ STAGE PLANS:
       Map Operator Tree:
           TableScan
             alias: t11_n1
-            filterExpr: (struct(id,lineid)) IN (const struct('1234-1111-0074578664','3'), const struct('1234-1111-0074578695',1)) (type: boolean)
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/test/results/clientpositive/structin.q.out:99]: Yes, some of these are just hidden by Hive implementation of the operator.
Bonus point because stricter typing on the parameters made the predicate vectorizable :)

File: ql/src/test/results/clientpositive/stat_estimate_related_col.q.out [hunk 15]
```diff
@@ -313,10 +313,10 @@ STAGE PLANS:
       Map Operator Tree:
           TableScan
             alias: t1
-            filterExpr: (((b = 2) or (b = 3)) and (b = 2)) (type: boolean)
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/test/results/clientpositive/stat_estimate_related_col.q.out:316]: It is the other way around: the OR/AND transformation does not happen anymore on the non CBO path. Note that the input query has an IN clause so it stays the same.

File: ql/src/test/results/clientpositive/spark/vector_between_in.q.out [hunk 16]
```diff
@@ -264,16 +264,16 @@ STAGE PLANS:
             Map Operator Tree:
                 TableScan
                   alias: decimal_date_test
-                  filterExpr: (cdecimal1) IN (2365.8945945946, 881.0135135135, -3367.6517567568) (type: boolean)
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/test/results/clientpositive/spark/vector_between_in.q.out:267]: Yes, @vineetgarg02  also noted this. I have created https://issues.apache.org/jira/browse/HIVE-23222 as a follow-up.

File: ql/src/test/results/clientpositive/perf/tez/constraints/cbo_query89.q.out [hunk 17]
```diff
@@ -77,7 +77,7 @@ HiveProject(i_category=[$0], i_class=[$1], i_brand=[$2], s_store_name=[$3], s_co
...
...E'Home':VARCHAR(2147483647) CHARACTER SET "UTF-16LE", _UTF-16LE'Books':VARCHAR(2147483647) CHARACTER SET "UTF-16LE", _UTF-16LE'Electronics':VARCHAR(2147483647) CHARACTER SET "UTF-16LE", _UTF-16LE'Shoes':VARCHAR(2147483647) CHARACTER SET "UTF-16LE", _UTF-16LE'Jewelry':VARCHAR(2147483647) CHARACTER SET "UTF-16LE", _UTF-16LE'Men':VARCHAR(2147483647) CHARACTER SET "UTF-16LE"), OR(AND(IN($12, _UTF-16LE'Home':VARCHAR(2147483647) CHARACTER SET "UTF-16LE", _UTF-16LE'Books':VARCHAR(2147483647) CHARACTER SET "UTF-16LE", _UTF-16LE'Electronics':VARCHAR(2147483647) CHARACTER SET "UTF-16LE"), IN($10, _UTF-16LE'wallpaper':VARCHAR(2147483647) CHARACTER SET "UTF-16LE", _UTF-16LE'parenting':VARCHAR(2147483647) CHARACTER SET "UTF-16LE", _UTF-16LE'musical':VARCHAR(2147483647) CHARACTER SET "UTF-16LE")), AND(IN($12, _UTF-16LE'Shoes':VARCHAR(2147483647) CHARACTER SET "UTF-16LE", _UTF-16LE'Jewelry':VARCHAR(2147483647) CHARACTER SET "UTF-16LE", _UTF-16LE'Men':VARCHAR(2147483647) CHARACTER SET "UTF-16LE"), IN($10, _UTF-16LE'womens':VARCHAR(2147483647) CHARACTER SET "UTF-16LE", _UTF-16LE'birdal':VARCHAR(2147483647) CHARACTER SET "UTF-16LE", _UTF-16LE'pants':VARCHAR(2147483647) CHARACTER SET "UTF-16LE"))))])
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/test/results/clientpositive/perf/tez/constraints/cbo_query89.q.out:80]: Yes, I agree. It does not make much sense. I have created https://issues.apache.org/jira/browse/HIVE-23224 to tackle this.

File: ql/src/test/results/clientpositive/perf/tez/constraints/cbo_query98.q.out [hunk 18]
```diff
@@ -75,7 +75,7 @@ HiveProject(i_item_desc=[$0], i_category=[$1], i_class=[$2], i_current_price=[$3
       HiveAggregate(group=[{1, 2, 3, 4, 5}], agg#0=[sum($8)])
         HiveJoin(condition=[=($7, $0)], joinType=[inner], algorithm=[none], cost=[not available])
           HiveProject(i_item_sk=[$0], i_item_id=[$1], i_item_desc=[$4], i_current_price=[$5], i_class=[$10], i_category=[$12])
-            HiveFilter(condition=[IN($12, _UTF-16LE'Jewelry', _UTF-16LE'Sports', _UTF-16LE'Books')])
+            HiveFilter(condition=[IN($12, _UTF-16LE'Jewelry':VARCHAR(2147483647) CHARACTER SET "UTF-16LE", _UTF-16LE'Sports':VARCHAR(2147483647) CHARACTER SET "UTF-16LE", _UTF-16LE'Books':VARCHAR(2147483647) CHARACTER SET "UTF-16LE")])
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/test/results/clientpositive/perf/tez/constraints/cbo_query98.q.out:78]: I'll tackle this as part of HIVE-23224.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/type/TypeCheckProcFactory.java [hunk 19]
```diff
@@ -1050,111 +1070,130 @@ protected T getXpathOrFuncExprNodeDesc(ASTNode expr,
...
+
+      if (exprFactory.isPOSITIVEFuncCallExpr(expr)) {
+        // UDFOPPositive is a no-op.
+        assert (exprFactory.getExprChildren(expr).size() == 1);
+        expr = exprFactory.getExprChildren(expr).get(0);
+      } else if (exprFactory.isNEGATIVEFuncCallExpr(expr)) {
+        // UDFOPNegative should always be folded.
+        assert (exprFactory.getExprChildren(expr).size() == 1);
+        T input = exprFactory.getExprChildren(expr).get(0);
+        if (exprFactory.isConstantExpr(input)) {
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/type/TypeCheckProcFactory.java:1105]: Only if `foldExpr` was true. Now for `+` and `-` we always do it in both paths. Otherwise I observed new failures when the UDFs are initialized (this is done to get their return type). The reason is that some of these UDFs are implemented in such a way that they verify that certain input parameters are constants, and they were getting a call to expression `UDFOPNegative` on top of a literal.

[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/type/TypeCheckProcFactory.java:1105]: `UDFOPPositive` is just above this block btw.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/type/TypeCheckProcFactory.java [hunk 20]
```diff
@@ -972,55 +1005,42 @@ protected T getXpathOrFuncExprNodeDesc(ASTNode expr,
...
-
-          if (hasNullValue) {
-            T nullConst = exprFactory.createConstantExpr(exprFactory.getTypeInfo(columnDesc), null);
-            if (outputOpList.size() == 0) {
-              // we have found only null values...remove the IN ; it will be null all the time.
-              return nullConst;
+          ListMultimap<TypeInfo, T> expressions = ArrayListMultimap.create();
+          for (int i = 1; i < children.size(); i++) {
+            T columnDesc = children.get(0);
+            T valueDesc = interpretNodeAsConstantStruct(columnDesc, children.get(i));
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/type/TypeCheckProcFactory.java:1011]: Yes, you are right. The problem is that the method name is very unfortunate. This can generate struct of constants, expressions, or anything else. I changed the method name to `interpretNode` and improve the documentation for it.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/type/RexNodeExprFactory.java [hunk 21]
```diff
@@ -0,0 +1,1021 @@
...
+    BigDecimal totalMonths = BigDecimal.valueOf(v.getTotalMonths());
+    return rexBuilder.makeIntervalLiteral(totalMonths,
+        new SqlIntervalQualifier(TimeUnit.YEAR, TimeUnit.MONTH, new SqlParserPos(1, 1)));
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected RexLiteral createIntervalMonthConstantExpr(String value) {
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/type/RexNodeExprFactory.java:498]: I think you are right that this can be refined further. To be honest, my main goal here was just to mimic existing interface (which was based on already existing calls) and separate concepts without refactoring more than necessary... and still look at this jumbo patch. However, I think you are right. Maybe the factory should be split into smaller classes... Or maybe we should create a single method and pass information as a parameter. I have created https://issues.apache.org/jira/browse/HIVE-23225.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/type/RexNodeExprFactory.java (hunk 7 above)
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/type/RexNodeExprFactory.java:835]: I am not sure... The method is utterly ugly because it mixes UDF with Calcite RexLiteral. That logic should simply not be in `TypeCheckProcFactory`. The reason why I had to leave it there is because we convert back CASE into COALESCE while going back from Calcite to Hive. However, I think from AST to Calcite this is not even needed. Hence, I am just going to return `false` for the `RexNodeExprFactory` implementation and cross my fingers...

Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/type/RexNodeExprFactory.java:835]: In the future, I hope we can get rid of that method all together.

Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/type/RexNodeExprFactory.java:835]: https://issues.apache.org/jira/browse/HIVE-23226

File: ql/src/java/org/apache/hadoop/hive/ql/parse/type/HiveFunctionHelper.java (hunk 10 above)
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/type/HiveFunctionHelper.java:236]: Tbh this was in my TODO list: many of those methods are not needed anymore in `RexConverter`. Hence my idea is to simply move them into `HiveFunctionHelper`. However, I did not want to do it as part of that patch to avoid further refactoring, as it was going to be shown as a bunch of removed code plus new code. I have created https://issues.apache.org/jira/browse/HIVE-23227 to tackle this.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/RexNodeConverter.java (hunk 11 above)
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/RexNodeConverter.java:681]: `value` cannot be null if we are here (note the check in L569). For the rest of types, sometimes there is a check on null value indeed, but it is because value is not final and it is replaced within those case blocks.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/type/TypeCheckProcFactory.java [hunk 22]
```diff
@@ -972,55 +1005,42 @@ protected T getXpathOrFuncExprNodeDesc(ASTNode expr,
...
-                genericUDF = new GenericUDFOPOr();
-                children.clear();
-                children.addAll(orOperands);
-              }
+          children.clear();
+          List<T> newExprs = new ArrayList<>();
+          int numEntries = expressions.keySet().size();
+          if (numEntries == 1) {
+            children.addAll(expressions.asMap().values().iterator().next());
+            funcText = "in";
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/type/TypeCheckProcFactory.java:1033]: I am going to add more information here because I think all this part was misleading, but here is the basic idea. We split the IN clause into different IN clauses, one for each different value type. As you know, Hive and Calcite treat types in IN clauses differently and it was practically impossible to find some correct implementation with a common part unless this was done. Then I rely on Calcite to merge the IN clauses if they are type compatible. This seems to be working fine.
I think this also answers your questions below.
1) We may have multiple values per type (observe it is a multimap). This is precisely what I want, to group those values by type.
2) In previous version, when IN clause had multiple null values, a single null value was added. Now I add multiple null values (basically I omitted the deduplication). I have added the deduplication back, although this can be done by Calcite.
3) For the twisted part, the initialization needs be done within the children processing because it depends on the type of the value itself, unless it is null in which case it depends on the column type.

Could this logic be improved or rewritten? Definitely... But I can tell you I work hard on this and the problem is that there are two many corner cases and two different type systems, and it is very easy that even the most minimal change produces incorrect results in any of them. That is why I came up with this solution to split IN clause depending on value type.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/type/RexNodeExprFactory.java [hunk 23]
```diff
@@ -0,0 +1,1021 @@
...
+      return res;
+    }
+    return ((RexLiteral) expr).getValue4();
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected String getConstantValueAsString(RexNode expr) {
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/type/RexNodeExprFactory.java:709]: This will be addressed in https://issues.apache.org/jira/browse/HIVE-23225.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/jdbc/JDBCExpandExpressionsRule.java (hunk 9 above)
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/jdbc/JDBCExpandExpressionsRule.java:200]: Tbh that method follows same contract as all other `RexConverter` methods. I see the value on what you are suggesting, since with a better contract, those methods could be easily called from within rules, etc. Since we will address the `RexConverter` refactoring in HIVE-23227, I have moved this method to `RexConverter` and I have left an additional reminder in the JIRA:
https://issues.apache.org/jira/browse/HIVE-23227?focusedCommentId=17085421&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17085421

File: ql/src/test/results/clientpositive/perf/tez/query82.q.out [hunk 24]
```diff
@@ -85,7 +85,7 @@ Stage-0
                               Select Operator [SEL_86] (rows=297 width=400)
                                 Output:["_col0","_col1","_col2","_col3"]
                                 Filter Operator [FIL_85] (rows=297 width=404)
-                                  predicate:(i_current_price BETWEEN 30 AND 60 and (i_manufact_id) IN (437, 129, 727, 663) and i_item_sk is not null)
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/test/results/clientpositive/perf/tez/query82.q.out:88]: I checked, I thought it was worth it. `i_current_price` min is 0.0900000035762787 and max 99.9899978637695; thus, assuming uniform distribution, BETWEEN will filter ~70% of the data. On the other hand, the min for `i_manufact_id` is 1 and the max is 1000; assuming the same, IN will filter >99% of the data.

File: ql/src/test/results/clientpositive/pcs.q.out [hunk 25]
```diff
@@ -1807,12 +1807,12 @@ STAGE PLANS:
       Map Operator Tree:
           TableScan
             alias: pcs_t1
-            filterExpr: ((key = 3) or (((((ds = '2000-04-08') or key is not null) and (key = 2)) or ((ds <> '2000-04-08') and (key = 3))) and ((key + 5) > 0))) (type: boolean)
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/test/results/clientpositive/pcs.q.out:1810]: Good catch. This is a missed optimization in Calcite. I am not sure why it is not happening now... I am wondering whether it is some weird interaction or maybe has to do with the typing of the constants. I have created https://issues.apache.org/jira/browse/HIVE-23228 to examine it.

File: ql/src/test/results/clientpositive/in_typecheck_char.q.out [hunk 26]
```diff
@@ -213,10 +214,10 @@ STAGE PLANS:
       Map Operator Tree:
           TableScan
             alias: ax
-            filterExpr: (((s = 'a') and (t = 'a         ')) or (null and (t = 'bb        '))) is null (type: boolean)
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply ql/src/test/results/clientpositive/in_typecheck_char.q.out:216]: Indeed old predicate seems more efficient. Probably this happened through the inference of the types for IN expression. I guess we made some predicates better (the one above) while this one got worse. I will leave it in this patch but it is worth taking a look at this. I created https://issues.apache.org/jira/browse/HIVE-23229.
